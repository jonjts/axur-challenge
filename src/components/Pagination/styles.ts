import styled from 'styled-components';

export const Container = styled.div`
`;

export const ListItem = styled.div`
    max-width: 100%;
    display: flex;
    flex: 1;
    justify-content: flex-start;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;  

`;

export const Row = styled.div`
    width: 90px;
    flex: 1;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    margin-top: ${({ theme: { spacings: { sm } } }) => sm.pixel};
`;

export const Footer = styled.div`
    display: flex;
    flex: 1;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    margin-top: ${({ theme: { spacings: { sm } } }) => sm.pixel};
`;