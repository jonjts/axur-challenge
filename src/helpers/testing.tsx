import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import ThemeProvider from '../providers/Theme';
import InspectionProvider from '../providers/Inspection'

export const putProviders = (component: JSX.Element) => (
    <ThemeProvider>
        <InspectionProvider>
            {component}
        </InspectionProvider>
    </ThemeProvider>
);

export const matchSnapshotWithProviders = (component: JSX.Element) => {
    const tree = renderer.create(putProviders(component)).toJSON();
    expect(tree).toMatchSnapshot();
    return tree;
};

export const renderWithProviders = (component: JSX.Element) => {
    return render(putProviders(component));
};
