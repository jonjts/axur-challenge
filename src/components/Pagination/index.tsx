import React, { useState, useEffect, memo, useMemo } from 'react';

import Button from '../Button';

import { Container, ListItem, Row, Footer } from './styles';

type Props = {
    items: Array<string>,
    onItemClicked(item: string): void,
    itemsPerPage?: number
};

const Pagination: React.FC<Props> = ({ items, onItemClicked, itemsPerPage = 10 }) => {

    const [itemsToShow, setItemsToShow] = useState(new Array<string>());
    const [page, setPage] = useState(1);
    const [disableNextPage, setDisableNextPage] = useState(true);
    const [disabledPreviousPage, setDisablePreviousPage] = useState(true);

    const totalPage = useMemo(() => {
        if (items.length) {
            const total = items.length / itemsPerPage;
            return (total >= 0 && total <= 1) ? 1 : total.toFixed();
        }
        return 1;
    }, [items, itemsPerPage]);

    useEffect(() => {
        setPage(1);
    }, [items]);

    useEffect(() => {
        const total = items.length - 1;
        let limit = page * itemsPerPage;
        let offset = limit - itemsPerPage;

        if (limit > total) {
            limit = total;
            setDisableNextPage(true);
        } else {
            setDisableNextPage(false);
        }

        if (page === 1) {
            setDisablePreviousPage(true);
        } else {
            setDisablePreviousPage(false);
        }

        if (offset < 0) {
            offset = 0;
        }

        setItemsToShow([...items.slice(offset, limit)]);
    }, [items, itemsPerPage, page]);

    return (
        <Container>
            {
                itemsToShow.map(item => (
                    <ListItem key={item}>
                        <Button
                            isLink
                            onClick={() => onItemClicked(item)}
                        >
                            {item}
                        </Button>
                    </ListItem>
                ))
            }
            <Footer data-testid={`${page}-${totalPage}`} >
                <label>{`PÃ¡gina ${page} de ${totalPage}`}</label>
                <Row>
                    <Button
                        disabled={disabledPreviousPage}
                        onClick={() => setPage(page - 1)}
                    >
                        {'<'}
                    </Button>
                    <Button
                        disabled={disableNextPage}
                        onClick={() => setPage(page + 1)}
                    >
                        {'>'}
                    </Button>
                </Row>
            </Footer>
        </Container>
    );
};

export default memo(Pagination);