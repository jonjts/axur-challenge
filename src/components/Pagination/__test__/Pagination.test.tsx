import { matchSnapshotWithProviders, renderWithProviders } from '../../../helpers/testing';

import Pagination from '../index';

const items = [
  'https://axur.com',
  'https://axur.com',
  'https://axur.com',
  'https://axur.com',
  'https://axur.com'
];

const getTotalPages = (itemsPerPage: number) => {
  if (items.length) {
    const total = items.length / itemsPerPage;
    return (total >= 0 && total <= 1) ? 1 : total.toFixed();
  }
  return 1;
};

describe('Button', () => {
  test('match with snapshot', () => {
    matchSnapshotWithProviders(<Pagination
      items={items}
      itemsPerPage={1}
      onItemClicked={() => { }}
    />);
  });

  test('renders correctly', () => {
    const result = renderWithProviders(<Pagination
      items={items}
      itemsPerPage={1}
      onItemClicked={() => { }}
    />);
    expect(result.getByTestId(`1-${getTotalPages(1)}`)).toBeInTheDocument();
  });
});
