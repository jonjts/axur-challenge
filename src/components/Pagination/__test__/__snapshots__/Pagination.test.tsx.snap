// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button match with snapshot 1`] = `
<div
  className="sc-hKgILt chrFSH"
>
  <div
    className="sc-gKsewC kLNfSF"
    data-testid="1-5"
  >
    <label>
      PÃ¡gina 1 de 5
    </label>
    <div
      className="sc-jSgupP fbhNlw"
    >
      <button
        className="sc-gsTCUz dPeWPU"
        data-testid="btn"
        onClick={[Function]}
      >
        &lt;
      </button>
      <button
        className="sc-gsTCUz dPeWPU"
        data-testid="btn"
        onClick={[Function]}
      >
        &gt;
      </button>
    </div>
  </div>
</div>
`;
